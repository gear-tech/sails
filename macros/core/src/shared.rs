use crate::route;
use proc_macro2::TokenStream as TokenStream2;
use proc_macro_error::abort;
use quote::{quote, ToTokens};
use std::collections::BTreeMap;
use syn::{
    spanned::Spanned, FnArg, Ident, ImplItem, ImplItemFn, ItemImpl, Pat, PathArguments, Receiver,
    ReturnType, Signature, Type, TypePath, TypeTuple, WhereClause,
};

/// A struct that represents the type of an `impl` block.
pub(crate) struct ImplType<'a> {
    path: &'a TypePath,
    args: &'a PathArguments,
    constraints: Option<&'a WhereClause>,
}

impl<'a> ImplType<'a> {
    pub(crate) fn new(item_impl: &'a ItemImpl) -> Self {
        let path = {
            let item_impl_type = item_impl.self_ty.as_ref();
            if let Type::Path(type_path) = item_impl_type {
                type_path
            } else {
                abort!(
                    item_impl_type.span(),
                    "Failed to parse impl type: {}",
                    item_impl_type.to_token_stream()
                )
            }
        };
        let args = &path.path.segments.last().unwrap().arguments;
        let constraints = item_impl.generics.where_clause.as_ref();
        Self {
            path,
            args,
            constraints,
        }
    }

    pub(crate) fn path(&self) -> &TypePath {
        self.path
    }

    pub(crate) fn args(&self) -> &PathArguments {
        self.args
    }

    pub(crate) fn constraints(&self) -> Option<&WhereClause> {
        self.constraints
    }
}

/// Represents parts of a handler function.
#[derive(Clone)]
pub(crate) struct Func<'a> {
    ident: &'a Ident,
    receiver: Option<&'a Receiver>,
    params: Vec<(&'a Ident, &'a Type)>,
    result: Type,
    is_async: bool,
}

impl<'a> Func<'a> {
    pub(crate) fn from(handler_signature: &'a Signature) -> Self {
        let func = &handler_signature.ident;
        let receiver = handler_signature.receiver();
        let params = Self::extract_params(handler_signature).collect();
        let result = Self::extract_result(handler_signature);
        Self {
            ident: func,
            receiver,
            params,
            result,
            is_async: handler_signature.asyncness.is_some(),
        }
    }

    pub(crate) fn ident(&self) -> &Ident {
        self.ident
    }

    pub(crate) fn receiver(&self) -> Option<&Receiver> {
        self.receiver
    }

    pub(crate) fn params(&self) -> &[(&Ident, &Type)] {
        &self.params
    }

    pub(crate) fn result(&self) -> &Type {
        &self.result
    }

    pub(crate) fn is_async(&self) -> bool {
        self.is_async
    }

    fn extract_params(handler_signature: &Signature) -> impl Iterator<Item = (&Ident, &Type)> {
        handler_signature.inputs.iter().filter_map(|arg| {
            if let FnArg::Typed(arg) = arg {
                let arg_ident = if let Pat::Ident(arg_ident) = arg.pat.as_ref() {
                    &arg_ident.ident
                } else {
                    abort!(arg.span(), "Unnamed arguments are not supported");
                };
                return Some((arg_ident, arg.ty.as_ref()));
            }
            None
        })
    }

    fn extract_result(handler_signature: &Signature) -> Type {
        match &handler_signature.output {
            ReturnType::Type(_, ty) => *ty.to_owned(),
            ReturnType::Default => Type::Tuple(TypeTuple {
                paren_token: Default::default(),
                elems: Default::default(),
            }),
        }
    }
}

pub(crate) fn discover_invocation_targets(
    item_impl: &ItemImpl,
    filter: impl Fn(&ImplItemFn) -> bool,
    allow_empty_route: bool, // Even though we always pass `false` here, we keep this parameter for the case when we want to allow anonymously exposed services
) -> BTreeMap<String, &Signature> {
    item_impl
        .items
        .iter()
        .filter_map(|item| {
            if let ImplItem::Fn(fn_item) = item {
                if filter(fn_item) {
                    let route = route::invocation_route(fn_item);
                    if route.1.is_empty() && !allow_empty_route {
                        abort!(route.0, "Empty route is not allowed")
                    }
                    return Some((route, &fn_item.sig));
                }
            }
            None
        })
        .fold(BTreeMap::new(), |mut result, (route, target)| {
            if let Some(duplicate) = result.insert(route.1, target) {
                abort!(
                    route.0,
                    "Route conflicts with one assigned to '{}'",
                    duplicate.ident.to_string()
                );
            }
            result
        })
}

pub(crate) fn generate_unexpected_input_panic(input_ident: &Ident, message: &str) -> TokenStream2 {
    let message_pattern = message.to_owned() + ": {}";
    quote!({
        let input = String::decode(&mut #input_ident)
            .unwrap_or_else(|_| {
                if #input_ident.len() <= 8 {
                    format!("0x{}", hex::encode(#input_ident))
                } else {
                    format!(
                        "0x{}..{}",
                        hex::encode(&#input_ident[..4]),
                        hex::encode(&#input_ident[#input_ident.len() - 4..]))
                }
            });
        panic!(#message_pattern, input)
    })
}
