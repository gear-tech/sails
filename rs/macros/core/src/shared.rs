use crate::route;
use proc_macro2::{Span, TokenStream as TokenStream2};
use proc_macro_error::abort;
use quote::{quote, ToTokens};
use std::collections::BTreeMap;
use syn::{
    punctuated::Punctuated, spanned::Spanned, FnArg, GenericArgument, Ident, ImplItem, ImplItemFn,
    ItemImpl, Lifetime, Pat, Path, PathArguments, PathSegment, Receiver, ReturnType, Signature,
    Token, Type, TypePath, TypeReference, TypeTuple, WhereClause,
};

pub(crate) fn impl_type(item_impl: &ItemImpl) -> (TypePath, PathArguments) {
    let item_impl_type = item_impl.self_ty.as_ref();
    let path = if let Type::Path(type_path) = item_impl_type {
        type_path.clone()
    } else {
        abort!(
            item_impl_type,
            "failed to parse impl type: {}",
            item_impl_type.to_token_stream()
        )
    };
    let args = path.path.segments.last().unwrap().arguments.clone();
    (path, args)
}

pub(crate) fn impl_constraints(item_impl: &ItemImpl) -> Option<WhereClause> {
    item_impl.generics.where_clause.clone()
}

/// Represents parts of a handler function.
#[derive(Clone)]
pub(crate) struct Func<'a> {
    ident: &'a Ident,
    receiver: Option<&'a Receiver>,
    params: Vec<(&'a Ident, &'a Type)>,
    result: Type,
    is_async: bool,
}

impl<'a> Func<'a> {
    pub(crate) fn from(handler_signature: &'a Signature) -> Self {
        let func = &handler_signature.ident;
        let receiver = handler_signature.receiver();
        let params = Self::extract_params(handler_signature).collect();
        let result = Self::extract_result(handler_signature);
        Self {
            ident: func,
            receiver,
            params,
            result,
            is_async: handler_signature.asyncness.is_some(),
        }
    }

    pub(crate) fn ident(&self) -> &Ident {
        self.ident
    }

    pub(crate) fn receiver(&self) -> Option<&Receiver> {
        self.receiver
    }

    pub(crate) fn params(&self) -> &[(&Ident, &Type)] {
        &self.params
    }

    pub(crate) fn result(&self) -> &Type {
        &self.result
    }

    pub(crate) fn is_async(&self) -> bool {
        self.is_async
    }

    fn extract_params(handler_signature: &Signature) -> impl Iterator<Item = (&Ident, &Type)> {
        handler_signature.inputs.iter().filter_map(|arg| {
            if let FnArg::Typed(arg) = arg {
                let arg_ident = if let Pat::Ident(arg_ident) = arg.pat.as_ref() {
                    &arg_ident.ident
                } else {
                    abort!(arg.span(), "unnamed arguments are not supported");
                };
                return Some((arg_ident, arg.ty.as_ref()));
            }
            None
        })
    }

    fn extract_result(handler_signature: &Signature) -> Type {
        result_type(handler_signature)
    }
}

pub(crate) fn result_type(handler_signature: &Signature) -> Type {
    match &handler_signature.output {
        ReturnType::Type(_, ty) => *ty.to_owned(),
        ReturnType::Default => Type::Tuple(TypeTuple {
            paren_token: Default::default(),
            elems: Default::default(),
        }),
    }
}

pub(crate) fn discover_invocation_targets(
    item_impl: &ItemImpl,
    filter: impl Fn(&ImplItemFn) -> bool,
) -> BTreeMap<String, (&ImplItemFn, usize)> {
    item_impl
        .items
        .iter()
        .enumerate()
        .filter_map(|item| {
            if let ImplItem::Fn(fn_item) = item.1 {
                if filter(fn_item) {
                    let route = route::invocation_route(fn_item);
                    return Some((route, (fn_item, item.0)));
                }
            }
            None
        })
        .fold(BTreeMap::new(), |mut result, (route, target)| {
            if let Some(duplicate) = result.insert(route.1, target) {
                abort!(
                    route.0,
                    "`route` attribute conflicts with one assigned to '{}'",
                    duplicate.0.sig.ident.to_string()
                );
            }
            result
        })
}

pub(crate) fn generate_unexpected_input_panic(
    input_ident: &Ident,
    message: &str,
    sails_path: &Path,
) -> TokenStream2 {
    let message_pattern = message.to_owned() + ": {}";
    let copy_ident = Ident::new(&format!("__{}", input_ident), Span::call_site());
    quote!({
        let mut #copy_ident = #input_ident;
        let input = String::decode(&mut #copy_ident)
            .unwrap_or_else(|_| {
                if #input_ident.len() <= 8 {
                    format!("0x{}", #sails_path::hex::encode(#input_ident))
                } else {
                    format!(
                        "0x{}..{}",
                        #sails_path::hex::encode(&#input_ident[..4]),
                        #sails_path::hex::encode(&#input_ident[#input_ident.len() - 4..]))
                }
            });
        panic!(#message_pattern, input)
    })
}

pub(crate) fn extract_lifetime_names(path_args: &PathArguments) -> Vec<String> {
    if let PathArguments::AngleBracketed(type_args) = path_args.clone() {
        type_args
            .args
            .into_iter()
            .filter_map(|a| {
                if let GenericArgument::Lifetime(lifetime) = a {
                    Some(lifetime.ident.to_string())
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
    } else {
        Vec::<String>::new()
    }
}

pub(crate) fn replace_any_lifetime_with_static(ty: Type) -> Type {
    match ty {
        Type::Reference(r) => {
            if r.lifetime.is_some() {
                Type::Reference(TypeReference {
                    and_token: r.and_token,
                    lifetime: Some(Lifetime::new("'static", Span::call_site())),
                    mutability: r.mutability,
                    elem: r.elem,
                })
            } else {
                Type::Reference(r)
            }
        }
        Type::Path(p) => Type::Path(TypePath {
            path: replace_lifetime_with_static_in_path(p.path),
            qself: p.qself,
        }),
        _ => ty,
    }
}

fn replace_lifetime_with_static_in_path(path: Path) -> Path {
    let mut segments: Punctuated<PathSegment, Token![::]> = Punctuated::new();
    for s in path.segments {
        segments.push(PathSegment {
            ident: s.ident,
            arguments: replace_lifetime_with_static_in_path_args(s.arguments),
        });
    }
    Path {
        leading_colon: path.leading_colon,
        segments,
    }
}

fn replace_lifetime_with_static_in_path_args(path_args: PathArguments) -> PathArguments {
    if let PathArguments::AngleBracketed(mut type_args) = path_args {
        type_args.args.iter_mut().for_each(|a| match a {
            GenericArgument::Lifetime(lifetime) => {
                *lifetime = Lifetime::new("'static", Span::call_site());
            }
            GenericArgument::Type(ty) => *ty = replace_any_lifetime_with_static(ty.clone()),
            _ => {}
        });
        PathArguments::AngleBracketed(type_args)
    } else {
        path_args
    }
}
