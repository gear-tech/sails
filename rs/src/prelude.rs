extern crate alloc;

pub use alloc::{
    borrow,
    borrow::ToOwned,
    boxed,
    boxed::Box,
    fmt, format, rc, str, string,
    string::{String, ToString},
    vec,
    vec::Vec,
};
pub use core::{
    any, array, ascii, assert_eq, assert_ne, cell, char, clone, cmp, convert, debug_assert,
    debug_assert_eq, debug_assert_ne, default, future, hash, hint, iter, marker, matches, mem, num,
    ops, option, panic, pin, prelude::rust_2021::*, primitive, ptr, result, slice, task, time,
    todo, unimplemented, unreachable, write, writeln,
};

/// Collection types.
///
/// See [`alloc::collections`] & [`hashbrown`].
pub mod collections {
    extern crate alloc;

    pub use ::hashbrown::{hash_map, hash_set, HashMap, HashSet};
    pub use alloc::collections::*;

    /// Reexports from [`hashbrown`].
    pub mod hashbrown {
        pub use ::hashbrown::{Equivalent, TryReserveError};
    }
}
/// Utilities related to FFI bindings.
///
/// See [`alloc::ffi`] & [`core::ffi`].
pub mod ffi {
    extern crate alloc;

    pub use alloc::ffi::{CString, FromVecWithNulError, IntoStringError, NulError};
    pub use core::ffi::*;
}

pub use crate::{
    gstd::{program, route, service},
    types::*,
};

pub use parity_scale_codec::{self as scale_codec, Decode, Encode, EncodeLike};
pub use scale_info::{self as scale_info, TypeInfo};
