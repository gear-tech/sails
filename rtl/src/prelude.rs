extern crate alloc;

pub use alloc::{
    borrow,
    borrow::ToOwned,
    boxed,
    boxed::Box,
    fmt, format, rc, str, string,
    string::{String, ToString},
    vec,
    vec::Vec,
};
pub use core::{
    any, array, ascii, assert_eq, assert_ne, cell, char, clone, cmp, convert, debug_assert,
    debug_assert_eq, debug_assert_ne, default, future, hash, hint, iter, marker, matches, mem, num,
    ops, option, panic, pin, prelude::rust_2021::*, primitive, ptr, result, slice, task, time,
    todo, unimplemented, unreachable, write, writeln,
};

/// Collection types.
///
/// See [`alloc::collections`] & [`hashbrown`].
///
/// [`alloc::collections`]: ::alloc::collections
pub mod collections {
    extern crate alloc;

    pub use ::hashbrown::{hash_map, hash_set, HashMap, HashSet};
    pub use alloc::collections::*;

    /// Reexports from [`hashbrown`].
    pub mod hashbrown {
        pub use ::hashbrown::{Equivalent, TryReserveError};
    }
}
/// Utilities related to FFI bindings.
///
/// See [`alloc::ffi`] & [`core::ffi`].
///
/// [`alloc::ffi`]: ::alloc::ffi
pub mod ffi {
    extern crate alloc;

    pub use alloc::ffi::*;
    pub use core::ffi::*;
}

pub use crate::types::*;

// Reexports from third-party libraries
pub use crate::scale::*;
