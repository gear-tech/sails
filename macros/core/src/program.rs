use crate::shared::{self, Func, ImplType};
use parity_scale_codec::Encode;
use proc_macro2::{Span, TokenStream as TokenStream2};
use proc_macro_error::abort;
use quote::quote;
use std::collections::BTreeMap;
use syn::{Ident, ItemImpl, Receiver, ReturnType, Signature, Type, TypePath, Visibility};

pub fn gprogram(program_impl_tokens: TokenStream2) -> TokenStream2 {
    let program_impl = syn::parse2(program_impl_tokens)
        .unwrap_or_else(|err| abort!(err.span(), "Failed to parse program impl: {}", err));
    let program_type = ImplType::new(&program_impl);
    let program_type_path = program_type.path();
    let program_type_args = program_type.args();
    let program_type_constraints = program_type.constraints();
    let program_ident = Ident::new("PROGRAM", Span::call_site());

    let (ctors_data, init) = generate_init(&program_impl, program_type_path, &program_ident);

    let (service_types, handle) = generate_handle(&program_impl, &program_ident);

    let services = service_types.map(|item| {
        let service_route = item.0;
        let service_type = item.1;
        quote!((#service_route, sails_idl_meta::AnyServiceMeta::new::<#service_type>()))
    });

    let ctors_params_structs = ctors_data.clone().map(|item| item.2);

    let ctors_meta_variants = ctors_data.map(|item| {
        let ctor_route = Ident::new(&item.0, Span::call_site());
        let ctor_params_struct_ident = item.1;
        quote!(#ctor_route(#ctor_params_struct_ident))
    });

    quote!(
        #program_impl

        impl #program_type_args sails_idl_meta::ProgramMeta for #program_type_path #program_type_constraints {
            fn constructors() -> scale_info::MetaType {
                scale_info::MetaType::new::<meta::ConstructorsMeta>()
            }

            fn services() -> impl Iterator<Item = (&'static str, sails_idl_meta::AnyServiceMeta)> {
                [
                    #(#services),*
                ].into_iter()
            }
        }

        use sails_rtl::prelude::Decode as __ProgramDecode;
        use sails_rtl::prelude::TypeInfo as __ProgramTypeInfo;

        #(#[derive(__ProgramDecode, __ProgramTypeInfo)] #ctors_params_structs )*

        mod meta {
            use super::*;

            #[derive(__ProgramTypeInfo)]
            pub enum ConstructorsMeta {
                #(#ctors_meta_variants),*
            }
        }

        #[cfg(target_arch = "wasm32")]
        pub mod wasm {
            use super::*;
            use sails_rtl::{gstd, hex, prelude::*};

            static mut #program_ident: Option<#program_type_path> = None;

            #init

            #handle
        }
    )
}

fn generate_init(
    program_impl: &ItemImpl,
    program_type_path: &TypePath,
    program_ident: &Ident,
) -> (
    impl Iterator<Item = (String, Ident, TokenStream2)> + Clone,
    TokenStream2,
) {
    let program_ctors = discover_program_ctors(program_impl, program_type_path);

    let input_ident = Ident::new("input", Span::call_site());

    let mut invocation_dispatches = Vec::with_capacity(program_ctors.len());
    let mut invocation_params_structs = Vec::with_capacity(program_ctors.len());

    for (invocation_route, program_ctor) in &program_ctors {
        let handler = Func::from(program_ctor);

        let invocation_params_struct_ident =
            Ident::new(&format!("__{}Params", invocation_route), Span::call_site());

        invocation_dispatches.push({
            let invocation_route_bytes = invocation_route.encode();
            let invocation_route_len = invocation_route_bytes.len();
            let handler_ident = handler.ident();
            let handler_await = handler.is_async().then(|| quote!(.await));
            let handler_args = handler.params().iter().map(|item| {
                let param_ident = item.0;
                quote!(request.#param_ident)
            });

            quote!(
                if #input_ident.starts_with(& [ #(#invocation_route_bytes),* ]) {
                    let request = #invocation_params_struct_ident::decode(&mut &#input_ident[#invocation_route_len..]).expect("Failed to decode request");
                    let program = #program_type_path :: #handler_ident (#(#handler_args),*) #handler_await;
                    static INVOCATION_ROUTE: [u8; #invocation_route_len] = [ #(#invocation_route_bytes),* ];
                    (program, INVOCATION_ROUTE.as_ref())
                }
            )
        });

        invocation_params_structs.push({
            let invocation_params_struct_members = handler.params().iter().map(|item| {
                let param_ident = item.0;
                let param_type = item.1;
                quote!(#param_ident: #param_type)
            });

            (
                invocation_route.clone(),
                invocation_params_struct_ident.clone(),
                quote!(
                    struct #invocation_params_struct_ident {
                        #(#invocation_params_struct_members),*
                    }
                ),
            )
        });
    }

    let init = if program_ctors.is_empty() {
        let unexpected_ctor_panic =
            shared::generate_unexpected_input_panic(&input_ident, "Unexpected ctor");

        quote!(
            #[no_mangle]
            extern "C" fn init() {
                let #input_ident = gstd::msg::load_bytes().expect("Failed to read input");
                if !#input_ident.is_empty() {
                    #unexpected_ctor_panic
                }
                unsafe {
                    #program_ident = Some(#program_type_path::default());
                }
                gstd::msg::reply_bytes(#input_ident, 0).expect("Failed to send output");
            }
        )
    } else {
        invocation_dispatches.push(shared::generate_unexpected_input_panic(
            &input_ident,
            "Unexpected ctor",
        ));

        quote!(
            #[gstd::async_init]
            async fn init() {
                let mut #input_ident: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
                let (program, invocation_route) = #(#invocation_dispatches)else*;
                unsafe {
                    #program_ident = Some(program);
                }
                gstd::msg::reply_bytes(invocation_route, 0).expect("Failed to send output");
            }
        )
    };

    (invocation_params_structs.into_iter(), init)
}

fn generate_handle(
    program_impl: &ItemImpl,
    program_ident: &Ident,
) -> (impl Iterator<Item = (String, Type)>, TokenStream2) {
    let service_ctors = discover_service_ctors(program_impl);

    let input_ident = Ident::new("input", Span::call_site());

    let mut invocation_dispatches = Vec::with_capacity(service_ctors.len());
    let mut last_resort_invocation_dispatch = None;
    let mut service_types = Vec::with_capacity(service_ctors.len());

    for (invocation_route, service_ctor) in &service_ctors {
        let service_ctor = Func::from(service_ctor);
        let service_ctor_ident = service_ctor.ident();

        service_types.push((invocation_route.clone(), service_ctor.result().clone()));

        if invocation_route.is_empty() {
            last_resort_invocation_dispatch = Some(quote!({
                    let program_ref = unsafe { #program_ident.as_ref() }.expect("Program not initialized");
                    let mut service = program_ref.#service_ctor_ident();
                    let output = service.handle(&#input_ident).await;
                    output
                }
            ));
        } else {
            invocation_dispatches.push({
                let invocation_route_bytes = invocation_route.encode();
                let invocation_route_len = invocation_route_bytes.len();

                quote!(
                    if #input_ident.starts_with(& [ #(#invocation_route_bytes),* ]) {
                        let program_ref = unsafe { #program_ident.as_ref() }.expect("Program not initialized");
                        let mut service = program_ref.#service_ctor_ident();
                        let output = service.handle(&#input_ident[#invocation_route_len..]).await;
                        static INVOCATION_ROUTE: [u8; #invocation_route_len] = [ #(#invocation_route_bytes),* ];
                        [INVOCATION_ROUTE.as_ref(), &output].concat()
                    }
                )
            });
        }
    }

    invocation_dispatches.push(last_resort_invocation_dispatch.unwrap_or_else(|| {
        shared::generate_unexpected_input_panic(&input_ident, "Unexpected service")
    }));

    (
        service_types.into_iter(),
        quote!(
            #[gstd::async_main]
            async fn main() {
                let mut #input_ident: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
                let output = #(#invocation_dispatches)else*;
                gstd::msg::reply_bytes(output, 0).expect("Failed to send output");
            }
        ),
    )
}

fn discover_program_ctors<'a>(
    program_impl: &'a ItemImpl,
    program_type_path: &'a TypePath,
) -> BTreeMap<String, &'a Signature> {
    let self_type_path = syn::parse_str::<TypePath>("Self").unwrap();
    shared::discover_invocation_targets(
        program_impl,
        |fn_item| {
            if matches!(fn_item.vis, Visibility::Public(_)) && fn_item.sig.receiver().is_none() {
                if let ReturnType::Type(_, output_type) = &fn_item.sig.output {
                    if let Type::Path(output_type_path) = output_type.as_ref() {
                        if output_type_path == &self_type_path
                            || output_type_path == program_type_path
                        {
                            return true;
                        }
                    }
                }
            }
            false
        },
        false,
    )
}

fn discover_service_ctors(program_impl: &ItemImpl) -> BTreeMap<String, &Signature> {
    shared::discover_invocation_targets(
        program_impl,
        |fn_item| {
            matches!(fn_item.vis, Visibility::Public(_))
                && matches!(
                    fn_item.sig.receiver(),
                    Some(Receiver {
                        mutability: None,
                        reference: Some(_),
                        ..
                    })
                )
                && fn_item.sig.inputs.len() == 1
                && !matches!(fn_item.sig.output, ReturnType::Default)
        },
        true,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn gprogram_discovers_public_associated_functions_returning_self_or_the_type_as_ctors() {
        let program_impl = syn::parse2(quote!(
            impl MyProgram {
                fn non_public_associated_func_returning_self() -> Self {}
                fn non_public_associated_func_returning_type() -> MyProgram {}
                fn non_public_associated_func_returning_smth() -> u32 {}
                pub fn public_associated_func_returning_self() -> Self {}
                pub fn public_associated_func_returning_type() -> MyProgram {}
                pub fn public_associated_func_returning_smth() -> u32 {}
                fn non_public_method_returning_self(&self) -> Self {}
                fn non_public_method_returning_type(&self) -> MyProgram {}
                fn non_public_method_returning_smth(&self) -> u32 {}
                pub fn public_method_returning_self(&self) -> Self {}
                pub fn public_method_returning_type(&self) -> MyProgram {}
                pub fn public_method_returning_smth(&self) -> u32 {}
            }
        ))
        .unwrap();
        let program_type_path = ImplType::new(&program_impl).path().clone();

        let discovered_ctors = discover_program_ctors(&program_impl, &program_type_path)
            .iter()
            .map(|s| s.1.ident.to_string())
            .collect::<Vec<_>>();

        assert_eq!(discovered_ctors.len(), 2);
        assert!(discovered_ctors.contains(&String::from("public_associated_func_returning_self")));
        assert!(discovered_ctors.contains(&String::from("public_associated_func_returning_type")));
    }

    #[test]
    fn gprogram_discovers_public_methods_with_self_ref_only_and_some_return_as_service_funcs() {
        let program_impl = syn::parse2(quote!(
            impl MyProgram {
                fn non_public_associated_func_returning_smth() -> u32 {}
                fn non_public_associated_func_returning_unit() {}
                pub fn public_associated_func_returning_smth() -> MyProgram {}
                pub fn public_associated_func_returning_unit() {}
                fn non_public_method_returning_smth(&self) -> u32 {}
                fn non_public_method_returning_unit(&self) {}
                pub fn public_method_returning_smth(&self) -> u32 {}
                pub fn public_method_returning_smth_with_other_params(&self, p1: u32) -> u32 {}
                pub fn public_methos_returning_smth_and_consuming_self(self) -> u32 {}
            }
        ))
        .unwrap();

        let discovered_services = discover_service_ctors(&program_impl)
            .iter()
            .map(|s| s.1.ident.to_string())
            .collect::<Vec<_>>();

        assert_eq!(discovered_services.len(), 1);
        assert!(discovered_services.contains(&String::from("public_method_returning_smth")));
    }
}
