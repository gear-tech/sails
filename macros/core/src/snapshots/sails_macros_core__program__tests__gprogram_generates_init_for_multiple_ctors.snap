---
source: macros/core/src/program.rs
expression: result
---
impl MyProgram {
    pub async fn new(p1: u32, p2: String) -> Self {
        Self { p1, p2 }
    }
    pub fn new2(p2: String, p1: u32) -> Self {
        Self { p1, p2 }
    }
}
use sails_rtl_gstd::Decode as InvocationParamsStructsDecode;
use sails_rtl_gstd::TypeInfo as InvocationParamsStructsTypeInfo;
#[derive(InvocationParamsStructsDecode, InvocationParamsStructsTypeInfo)]
struct __NewParams {
    p1: u32,
    p2: String,
}
#[derive(InvocationParamsStructsDecode, InvocationParamsStructsTypeInfo)]
struct __New2Params {
    p2: String,
    p1: u32,
}
#[cfg(target_arch = "wasm32")]
pub mod wasm {
    use super::*;
    use sails_rtl_gstd::{*, gstd};
    pub(crate) static mut PROGRAM: Option<MyProgram> = None;
    #[gstd::async_init]
    async fn init() {
        let mut input: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
        let (program, invocation_route) = if input
            .starts_with(&[12u8, 78u8, 101u8, 119u8])
        {
            let request = __NewParams::decode(&mut &input[4usize..])
                .expect("Failed to decode request");
            let program = MyProgram::new(request.p1, request.p2).await;
            static INVOCATION_ROUTE: [u8; 4usize] = [12u8, 78u8, 101u8, 119u8];
            (program, INVOCATION_ROUTE.as_ref())
        } else if input.starts_with(&[16u8, 78u8, 101u8, 119u8, 50u8]) {
            let request = __New2Params::decode(&mut &input[5usize..])
                .expect("Failed to decode request");
            let program = MyProgram::new2(request.p2, request.p1);
            static INVOCATION_ROUTE: [u8; 5usize] = [16u8, 78u8, 101u8, 119u8, 50u8];
            (program, INVOCATION_ROUTE.as_ref())
        } else {
            let invocation_route = String::decode(&mut input)
                .expect("Failed to decode invocation route");
            panic!("Unknown init request: {}", invocation_route);
        };
        unsafe {
            PROGRAM = Some(program);
        }
        gstd::msg::reply_bytes(invocation_route, 0).expect("Failed to send output");
    }
}

